components/suggestion/command_provider/app_command_parser.ts

```diff
@@ -231,229 +739,155 @@ export class AppCommandParser {
+        const missing: AppField[] = [];
 
-        const position = tokens.length;
-        if (!lastToken) {
-            return [{suggestion: 'Could not find last token'}];
+        const values = parsed.values || [];
+        const fields = form.fields || [];
+        for (const field of fields) {
+            if (field.is_required && !values[field.name]) {
+                missing.push(field);
+            }
         }
 
-        const positionalField = form.fields.find((f) => f.position === position + 1);
-        if (positionalField) {
-            return this.getSuggestionsForField(positionalField, lastToken.value, pretext);
-        }
+        return missing;
+    }
 
-        const previousToken = tokens[position - 1];
-        if (previousToken && previousToken.type === 'flag') {
-            const flagField = form.fields.find((a) => a.name === previousToken.name);
-            if (flagField) {
-                return this.getSuggestionsForField(flagField, lastToken.value, pretext);
-            }
+    // getFlagNameSuggestions returns suggestions for flag names
+    getFlagNameSuggestions = (parsed: ParsedCommand): AutocompleteSuggestion[] => {
+        if (!parsed.form || !parsed.form.fields || !parsed.form.fields.length) {
             return [];
         }
 
-        // '', '-', or '--' (show named args)
-        if ('--'.startsWith(lastToken.value.trim())) {
-            const available = form.fields.filter((arg) => {
-                if (arg.position) {
-                    return false;
-                }
-                if (tokens.find((t) => t.type === 'flag' && t.name === arg.name)) {
-                    return false;
-                }
-                if (arg.name.startsWith(lastToken.name)) {
-                    return true;
-                }
+        // There have been 0 to 2 dashes in the command prior to this call, adjust.
+        let prefix = '--';
+        for (let i = parsed.incompleteStart - 1; i > 0 && i >= parsed.incompleteStart - 2 && parsed.command[i] === '-'; i--) {
+            prefix = prefix.substring(1);
+        }
 
-                return false;
+        const applicable = parsed.form.fields.filter((field) => field.label && field.label.startsWith(parsed.incomplete));
+        if (applicable) {
+            return applicable.map((f) => {
+                let suffix = '';
+                if (f.type === AppFieldTypes.USER) {
+                    suffix = ' @';
+                } else if (f.type === AppFieldTypes.CHANNEL) {
+                    suffix = ' ~';
+                }
+                return {
+                    complete: prefix + (f.label || f.name) + suffix,
+                    suggestion: '--' + (f.label || f.name),
+                    description: f.description,
+                    hint: f.hint,
+                };
             });
-            return this.getSuggestionsFromFlags(available);
         }
 
         return [{suggestion: 'Could not find any suggestions'}];
```

Comment:
Wonder if there is any case where this suggestion would appear along a "execute" suggestion or other kind of suggestions.